txpipe/photoz_stack.py

```diff
@@ -1,496 +1,282 @@
+                        pdfs_chunk = qp_bin.gen_obj.pdfs
+                    elif pdf_type == "interp":
+                        pdfs_chunk = qp_bin.gen_obj.yvals
+                    else:
+                        raise ValueError(f"TXPipe cannot yet use QP PDF type {pdf_type}")
+
+                    if weights is None:
+                        chunk_stack = pdfs_chunk.sum(axis=0)
+                        pdfs[i] += chunk_stack
+                        total_weight[i] += qp_bin.npdf
+                        # 2D bin
+                        pdfs[-1] += chunk_stack
+                        total_weight[-1] += qp_bin.npdf
+                    else:
+                        # This is not yet tested!
+                        chunk_stack = weights[sel] @ pdfs_chunk
+                        w = weights[sel].sum()
+                        pdfs[i] += chunk_stack
+                        total_weight[i] += w
+                        pdfs[-1] += chunk_stack
+                        total_weight[-1] += w
+                
+                
+        
+        # Collect the results from all processors
+        in_place_reduce(pdfs, self.comm)
+        in_place_reduce(total_weight, self.comm)
+        
+        # only root saves output
+        if self.rank == 0:
+            pdfs /= total_weight[:, None]
 
-    def get_metadata(self):
-        """
-        Load the z column and the number of bins
+            # Create a qp object for the n(z) information that we have.
+            if pdf_type == "interp":
+                q = qp.Ensemble(qp.interp, data={"xvals":z, "yvals":pdfs})
+            elif pdf_type == "hist":
+                q = qp.Ensemble(qp.hist, data={"bins":z, "pdfs":pdfs.T})
+            else:
+                raise ValueError(f"TXPipe cannot yet use QP PDF type {pdf_type}")
 
-        Returns
-        -------
-        z: array
-            Redshift column for photo-z PDFs
-        nbin:
-            Number of different redshift bins
-            to split into.
+            with self.open_output("photoz_stack", "w") as f:
+                f.write_ensemble(q)
 
-        """
-        # It's a bit odd but we will just get this from the file and
-        # then close it again, because we're going to use the
-        # built-in iterator method to get the rest of the data
 
-        # open_input is a method defined on the superclass.
-        # it knows about different file formats (pdf, fits, etc)
-        with self.open_input("source_photoz_pdfs") as photoz_file:
-            # This is the syntax for reading a complete HDF column
-            z = photoz_file["meta/xvals"][0, :]
 
-        # Save again but for the number of bins in the tomography catalog
-        with self.open_input("shear_tomography_catalog") as tomo_file:
-            nbin_source = tomo_file["tomography"].attrs["nbin"]
 
-        return z, nbin_source
 
 
-class TXPhotozLensStack(TXPhotozSourceStack):
+class TXTruePhotozStack(PipelineStage):
```

Comment:
is there a version where we are using e.g. a z_mc value? that would be essentially the same as this class but histograming a different column (though I guess qp does not have that option now, and maybe not worth having...?)

Reply:
I'm adding a config option to let people choose a different column here.  Using QP to generate draws would have to be from a different class though as the input would need to be a a QP file instead of a plain catalog. We can do that if needed though.